nestkernel/input_backend_mpi.cpp

```diff
@@ -0,0 +1,295 @@
+  }
+}
+
+void
+nest::InputBackendMPI::cleanup()
+{
+  // Disconnect all the MPI connection and send information about this disconnection
+  // Clean all the elements in the map
+  thread thread_id = kernel().vp_manager.get_thread_id();
+  // WARNING can be a bug if all threads send ending MPI connection and
+  // disconnect MPI message
+  for ( auto& it_comm : commMap_[ thread_id ] )
+  {
+    int value[ 1 ];
+    value[ 0 ] = thread_id;
+    MPI_Send( value, 1, MPI_INT, 0, 2, *it_comm.second.first );
+    MPI_Comm_disconnect( it_comm.second.first );
+    delete ( it_comm.second.first );
+  }
+  // clear map of devices
+  commMap_[ thread_id ].clear();
+  for ( auto& it_device : devices_[ thread_id ] )
+  {
+    it_device.second.first = nullptr;
+  }
+}
+
+void
+nest::InputBackendMPI::check_device_status( const DictionaryDatum& params ) const
+{
+  // nothing to do
+}
+
+void
+nest::InputBackendMPI::get_device_defaults( DictionaryDatum& params ) const
+{
+  // nothing to do
+}
+
+void
+nest::InputBackendMPI::get_device_status( const nest::InputDevice& device, DictionaryDatum& params_dictionary ) const
+{
+  // nothing to do
+}
+
+
+void
+nest::InputBackendMPI::get_status( lockPTRDatum< Dictionary, &SLIInterpreter::Dictionarytype >& ) const
+{
+  // nothing to do
+}
+
+void
+nest::InputBackendMPI::set_status( const DictionaryDatum& d )
+{
+  // nothing to do
+}
+
+
+void
+nest::InputBackendMPI::get_port( InputDevice* device, std::string* port_name )
+{
+  get_port( device->get_node_id(), device->get_label(), port_name );
+}
+
+void
+nest::InputBackendMPI::get_port( const index index_node, const std::string& label, std::string* port_name )
+{
+  // path of the file : path+label+id+.tx
+  // (file contains only one line with name of the port)
+  std::ostringstream basename;
+  const std::string& path = kernel().io_manager.get_data_path();
+  if ( not path.empty() )
+  {
+    basename << path << '/';
+  }
+  basename << kernel().io_manager.get_data_prefix();
+
+  if ( not label.empty() )
+  {
+    basename << label;
+  }
+  else
+  {
+    throw MPIFilePortsUnknown( index_node );
+  }
+  char add_path[ 150 ];
+  sprintf( add_path, "/%zu.txt", index_node );
+  basename << add_path;
+  std::cout << basename.rdbuf() << std::endl;
+  std::ifstream file( basename.str() );
+
+  if ( file.is_open() )
+  {
+    getline( file, *port_name );
+  }
+  file.close();
+}
+
+void
+nest::InputBackendMPI::receive_spike_train( const MPI_Comm& comm, InputDevice& device )
+{
+  // Send the first message with id of device and thread id
+  int message[ 2 ];
+  message[ 0 ] = device.get_node_id();
+  message[ 1 ] = kernel().vp_manager.get_thread_id();
+  MPI_Status status_mpi;
+  MPI_Send( message, 2, MPI_INT, 0, 0, comm );
+  // Receive the size of data
+  int shape[ 1 ];
+  MPI_Recv( &shape, 1, MPI_INT, MPI_ANY_SOURCE, message[ 1 ], comm, &status_mpi );
+  // Receive the data ( for the moment only spike time )
+  double spikes[ shape[ 0 ] ];
```

Comment:
My compiler gives a warning that using variable length arrays is a C99 feature. As we have switched to C++11, we should also be using C11, I think. However, this is not specified when compiling. I can create an issue on this.

Reply:
I choice variable length array because it's impossible to know the number of spike receiving before.
For avoid this, I need to modify the protocol of communication in order to define a fix number of spike receiving. This modifiication will complex the communication and can slow it but it will increase the robustness of the code.
What do you prefer?

Reply:
I would prefer avoiding variable length arrays. And I see that you have already changed to dynamic allocation, which should do the trick.